### ISSUE TITLE
azure-identity: Connection reset with InteractiveBrowserCredential in container with port binding for redirect URI

### ISSUE THREAD

**Issue by elisiariocouto** [Azure/azure-sdk-for-python#32400]
- **Package Name**: azure-identity
- **Package Version**: 1.14.0
- **Operating System**: macOS using Docker
- **Python Version**: 3.10

**Describe the bug**
Inside a container, I'm using `InteractiveBrowserCredential` with an App with `http://localhost:8000` as a valid callback URI. Forcing that URI in the `InteractiveBrowserCredential` to be that callback URI and binding the port 8000 on the host to the 8000 on the container, azure-identity detects that no browser is present in the container and gives me a link to call from the host machine. When I make that request from the host, I get no answer from the lib. If I make the request inside the container (using curl for instance), the authentication succeeds.

**To Reproduce**
Run this in a container with `docker run --rm -it -p 8000:8000 python:3.10-slim bash`

```python
from azure.identity import InteractiveBrowserCredential

credential = InteractiveBrowserCredential(
        tenant_id=<REDACTED>,
        client_id=<REDACTED>,
        redirect_uri="http://localhost:8000",
    )
record = credential.authenticate(scopes=[<REDACTED>])
```

In Safari, I get "the server unexpectedly dropped the connection".

**Comment by annatisch** [comment]
Thanks for the report @elisiariocouto!
We will investigate ASAP :)

**Comment by xiangyan99** [comment]
@rayluo , could you help to take a look?

**Comment by elisiariocouto** [comment]
I want to add that the `--net=host` does not work since the host is the Docker VM using macOS. Also, tried `http://0.0.0.0:8000` which binds successfully (?) but got the same issue.

**Comment by xiangyan99** [comment]
Unfortunately, msal does not support this scenario yet.

**Comment by elisiariocouto** [comment]
/unresolve

What is the reason for the connection drop? Do you accept MRs to address this?

**Comment by rayluo** [comment]
@elisiariocouto , can you try somehow upgrading your docker image with this experimental MSAL package and see if it helps? Your test will give us confidence on this enhancement.

    pip install --force-reinstall git+https://github.com/AzureAD/microsoft-authentication-library-for-python.git@docker-support

**Comment by elisiariocouto** [comment]
Same issue @rayluo.
I managed to do some troubleshooting, it appears that the problem is that `msal` is binding to 127.0.0.1 and it should bind to all the interfaces. I saw that you introduced a check on the `docker-support` branch but the check is returning `False` in my environment. It appears that there is no silver bullet, but if you guys want to support docker, the only way that I found is to check the existence of the `/.dockerenv` file.

**Comment by rayluo** [comment]
Thanks for your test, @elisiariocouto . Now we added the `/.dockerenv` check as a fall back. It shall work for docker on Mac. Mind test it one more time?

**Comment by elisiariocouto** [comment]
Worked like a charm. Thank you @rayluo for the fix. Do you know when is this going to be released?

**Comment by rayluo** [comment]
@elisiariocouto , thanks for the confirmation. We will merge in the fix soon.

We do not currently have a specific timeline for the next release. You can, however, click the "watch" button near the top of [MSAL repo](https://github.com/AzureAD/microsoft-authentication-library-for-python) to subscribe its release notification.

**Comment by joshfree** [comment]
@xiangyan99 does the version of MSAL that you're picking up for the November Azure Identity for Python SDK release include this fix?

**Comment by rayluo** [comment]
> @xiangyan99 does the version of MSAL that you're picking up for the November Azure Identity for Python SDK release include this fix?

Yes, it is shipped in the latest MSAL Python 1.25.
